### PR TITLE
Add bounds to ranges (limits pan and zoom)

### LINKED ISSUE #2942: Feature request: Constrain zoom out to fixed range/domain
On Twitter @nfaggian asked about being able to constrain the zoom tool to within a fixed range/domain on a plot. It sounds like a reasonable feature to me.

### PR DESCRIPTION
**Pull request by birdsarah**
Fixes #2942

### CODE REVIEW

File: bokeh/models/ranges.py [hunk 1]
```diff
@@ -90,6 +100,19 @@ class DataRange1d(DataRange):
     automatically computed end value.
     """)
 
+    bound_lower = Float(help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data. If not provided, will
+    default to the min of the data.
+    """)
```
Comment (bryevdv) [inline bokeh/models/ranges.py]: I think this is not quite adequate, I would suggest `None` means no bound, `"auto"` means default to the min of the data, and explicit values are explicit bounds. So this would be `Either(Auto, Float)`

File: bokeh/models/ranges.py [hunk 2]
```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
...
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```
Comment (bryevdv) [inline bokeh/models/ranges.py]: I'm not 100% certain this functionality is useful for categorical ranges, I think `upper` and `lower` imply an ordering, which is not the case here.

File: bokeh/models/ranges.py [hunk 3]
```diff
@@ -42,6 +42,16 @@ class Range1d(Range):
     The end of the range.
     """)
 
+    bound_lower = Either(Float, Datetime, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```
Comment (bryevdv) [inline bokeh/models/ranges.py]: The `Datetime` in here really bugs me. I'm not sure what a better solution would be offhand.

File: bokehjs/src/coffee/common/plot.coffee [hunk 4]
```diff
@@ -196,20 +196,33 @@ class PlotView extends ContinuumView
...
-        rng.set(range_info.xrs[name])
-        rng.get('callback')?.execute(@model)
+      @_update_single_range(rng, range_info.xrs[name], is_panning)
     for name, rng of @frame.get('y_ranges')
-      if rng.get('start') != range_info.yrs[name]['start'] or
-          rng.get('end') != range_info.yrs[name]['end']
-        rng.set(range_info.yrs[name])
-        rng.get('callback')?.execute(@model)
+      @_update_single_range(rng, range_info.yrs[name], is_panning)
     @unpause()
```
Comment (bryevdv) [inline bokehjs/src/coffee/common/plot.coffee]: This just needs to be tested with multiple ranges/axes. It might work just fine, but I'd like to see it demonstrated.

File: bokeh/models/ranges.py (hunk 2 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: Hi, I'm definitely happy to change the concepts/approach, but I think the end result is really helpful - I find it super weird that our default charts let you pan a bar chart, heat map beyond their max and min. On continuous axes you might be outside the data, but at least it has meaning on a categorical axis it has no meaning. But, for example, on a heat map you do still want panning and zooming, just not beyond the extents. 

I was tempted to implement, that if you don't set one, it gets set to `bound_lower=factors[0]` and `bound_upper=factors[-1]`

Just to re-state, I'm not disagreeing that `upper` and `lower` are conceptually odd, but I'd love to see us enable something like this, and perhaps push it further than this PR does.

File: bokeh/models/ranges.py (hunk 3 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: I copied the types from `min`. Are we in agreement that `bound_lower` should accept the same as `min`?

File: bokeh/models/ranges.py (hunk 1 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: I disagree, I think this is one of the cases people who've been asking for this feature want to see improve.

My interpretation was that if you do the following:
```python
p = figure()
p.scatter(lots of numbers)
```
Then your plot is auto-ranged (start and end are set to min and max of your data) and if you're zooming out you can't go off into infinity.

 I could be wrong, but I think its worth at least asking around.

I think people are also looking for the converse, that you can't zoom in crazy far either and this PR doesn't handle that use case.

File: bokehjs/src/coffee/common/plot.coffee (hunk 4 above)
Comment (birdsarah) [inline bokehjs/src/coffee/common/plot.coffee]: Added example to demonstrate multiple ranges. It works as you'd hope.

File: bokeh/models/ranges.py (hunk 1 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: I just realized I might have misunderstood your comment.

I got the type from `min`. 

Are you saying that the accepted values for `min`, `max`, `bound_lower`, and `bound_upper` on `DataRange1d` should be `Either(Auto, Float)`.

File: bokeh/models/ranges.py (hunk 3 above)
Comment (bryevdv) [inline bokeh/models/ranges.py]: Yes. I had forgotten it was already so loosely specified. Fixing that is for another day.

File: bokeh/models/ranges.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/models/ranges.py]: Can we leave it for a separate PR then? We could investigate what other projects do, e.g. before.

File: bokeh/models/ranges.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/models/ranges.py]: Well, I was suggesting that for `bound_lower` and `bound_upper`.

**Comment by birdsarah** [comment]
@bryevdv I think I got all your points. @brendancol want to take a look?

File: bokeh/models/ranges.py (hunk 2 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: Well, I was thinking that we should have consistent naming across the ranges, as then once you've learned how to do it for one range, then you don't have to look up how to do it for a different one. Maybe we can pick something that works for all?

---
d3 defines a rangeExtent and scaleExtent that would be used in this context.

https://github.com/mbostock/d3/wiki/Ordinal-Scales#ordinal_rangeExtent
`ordinal.rangeExtent()`
Returns a two-element array representing the extent of the scale's range, i.e., the smallest and largest values.

https://github.com/mbostock/d3/wiki/zoom-behavior#scaleExtent
`zoom.scaleExtent([extent])`
Specifies the zoom scale's allowed range as a two-element array, [minimum, maximum]. If not specified, returns the current scale extent, which defaults to [0, Infinity].

---

I also had a poke around Grammar of Graphics and didn't see anything conclusive, but it's a big book so may have missed it.

---

ggplot2 - http://docs.ggplot2.org/current/xylim.html
has the concept of xlim and ylim
xlim(15, 20) 
for discrete data you can do: xlim("a", "b", "c")

---

vega has a concept of domainMin/Max and rangeMin/Max domainMin/Max is for quantitative scales only, but rangeMin/Max appears to be for all (I haven't seen an example with this in action though so not 100%)

https://github.com/vega/vega/wiki/Scales

---

plotly - couldn't find it

---

In general, looking around I'm not finding a clear message, but I may be looking in the wrong place. 

Maybe @pzwang has opinions about this?

File: bokeh/models/ranges.py (hunk 1 above)
Comment (birdsarah) [inline bokeh/models/ranges.py]: > There's no sense which "no min" make sense to me, e.g. (there's always a min, sometimes it is computed), so I would not apply this suggestion to min

Hi @bryevdv, I'm confused. can you explain again?

To my mind: `DataRange1d` `start` looks like this:

````python
    start = Float(help="")
````
If you provide no start then the value is computed.

Similarly, I am proposing:
````python
    bound_lower = Float(help="")
````
and if you provide no bound_lower then the value is computed.

You said:
> I would suggest None means no bound, "auto" means default to the min of the data, and explicit values are explicit bounds. So this would be Either(Auto, Float)

Why does `None` mean "no bound" in the case of `bound_lower` but "please compute" in the case of `start`?

Comment (bryevdv) [inline bokeh/models/ranges.py]: Because `start` always has to be *something*, regardless of whether I supply that something, or whether that something gets computed. There is no case it which a start value literally not existing makes sense. Now, before there was only start, and making `start=None` as a way of saying "hey I'm not giving you a start value, figure it out yourself" seemed perfectly reasonable. In this context `None` is being used in the fairly common manner of `foo=None` keywoard arg that is computed if the user omits supplying it.

`lower_bound` by contrast has more "states", and we have to have a way to express all of them: 
* explicit user bounds 
* autocomputed bounds
* unbounded. 

Explicit is when a value is set, I think "auto" makes sense for "autocompute", that leaves `None`  for "unbounded", i.e. no bound, which also seems fairly sensible to me. Of course in this case `None` really means "None", not just a placeholder for a missing default. 

`start` and `lower_bound` sort of both "look like numbers" but they are different things, it's worth considering. `start` is a number, computed or given. `lower_bound` is a *policy* that describes how to do the computing for whenever `start` is computed. 

BTW we should document the precedence. What happens if a user sets `lower_bound` higher than the data min, who wins?

File: bokeh/models/ranges.py (hunk 2 above)
Comment (pzwang) [inline bokeh/models/ranges.py]: I agree with Sarah that allowing panning beyond the data range is undesirable behavior in most cases.  However, it should be noted that we oftentimes will overlay multiple datasets and plots onto the same plot area, in which case we will want to be able to pan/zoom beyond the boundary of any individual plot.

Bryan is also correct that having "lower" and "upper" as bounds implies an ordering *at the data level*, which does not necessarily exist for sets of factors.  Stated differently, even with these parameters available, without also providing the user some understanding of how their unordered data will be sorted, then they can't reason about what it means for two random elements from the set to be "min" and "max".

However, it should be noted here that at some point we *do* splat things onto the screen, and there is a draw order imposed at that time (or by that time).  In that context, then, it does make sense to refer to the leftmost/rightmost/bottommost/topmost data element.

I like the suggestion that really what we're trying to express is "bounds policy", and "min"/"max" is a shorthand for a particular "bounds policy" or "fit policy", which is "Limit to tight fit at a precise literal values", with a tuple expressing min and max.  For unordered sets, one could imagine a bounds policy that is just "Limit to tight data fit", or "Limit to tight data fit, with a padding of 10 pixels".  Alternatively, one can imagine bounds policy that is an offset: "Limit upper bound to maximum data data; lower bound defaults to offset of 12.3 data units".

Where is that #MidnightIdeation tag... Oh, found it.

Comment (bryevdv) [inline bokeh/models/ranges.py]: > for discrete data you can do: xlim("a", "b", "c")

I believe this is the correct answer for categorical data. Instead of two min/max properties, you provide the subset of categories that you want.

Comment (birdsarah) [inline bokeh/models/ranges.py]: Thanks for all this...some thoughts are brewing.

I wanted a consistent spelling across the ranges, and bokeh already has a notion of pre/suf-fixed things like `hover_line_alpha`,  `background_fill_alpha` etc.

What about a property with the suffix `_bound` or `_limit` on each of the things that end up turning into start end. So concretely:

 * Range1d
   * start - Either(Float, Datetime, Int)
   * start_bound - Either(Float, Datetime, Int, Auto, None, default=??)
   * end - Either(Float, Datetime, Int)
   * end_bound - Either(Float, Datetime, Int, Auto, None, default=??)
 * DataRange1d
  * start - Float
  * start_bound - Either(Float, Auto, None, default=Auto)
  * end - Float
  * end_bound - Either(Float, Auto, None, default=Auto)
 * FactorRange
  * factors = Either(List(String), List(Int))
  * factors_bound = Either(List(String), List(Int), Auto, None, default=Auto)

* Auto will just compute it to what your initial data is
* None will be no bound

This only leaves one question: If `factors=['bat', 'cat', 'dat', 'eat', 'fat', 'gat']` and your `factors_bound=['fat', 'cat', 'eat']`  - what is expected behavior - I am not sure. 

Ordered-like behavior: Take factors as the canonical order, then sort factors_bound into the same order and then take the uppermost and lower most elements. (As @pzwang said, factors end up in drawn-space eventually so this was inspired by that). That would leave a plot that shows `factors=['cat', 'dat', 'eat', 'fat']` and you can't pan outside that

xlim like behavior: Throw out the rows that aren't in the bound. That would leave a plot that shows `factors=['cat', 'eat', 'fat']` and you can't pan outside that.

The xlim behavior seems more intuitive to me, although I'm not sure why someone would use it. and not just set factors to what they want and use Auto for the bound.

Comment (bryevdv) [inline bokeh/models/ranges.py]: I had in mind `xlim` behavior, i.e., `factors_bound=['fat', 'cat', 'eat']` shows just those and nothing more. I imagine this kind of bound could be useful in a callback to subset the data without having to reload it?

I'm not sure about `start_bound` and `end_bound`. Remember, it is possible for `start > end` (i.e., if you want "reversed" range displayed) so I think `min` and `max` more clearly express the intent. Then the internals can sort out if `start` or `end` is the min, and do whatever is appropriate. Otherwise we force this on users to know and do the right thing and that seems fragile.

Comment (birdsarah) [inline bokeh/models/ranges.py]: Above I wrote

* Range1d
   * start - Either(Float, Datetime, Int)
   * start_bound - Either(Float, Datetime, Int, Auto, None, default=??)
   * end - Either(Float, Datetime, Int)
   * end_bound - Either(Float, Datetime, Int, Auto, None, default=??)

But I'd like to ammend that to 

* Range1d
   * start - Either(Float, Datetime, Int)
   * start_bound - Either(Float, Datetime, Int, Auto, None, default=Auto)
   * end - Either(Float, Datetime, Int)
   * end_bound - Either(Float, Datetime, Int, Auto, None, default=Auto)

Having `default=Auto` will lead to a change in behavior from what users are currently used to with Bokeh. However, I think this is ok because:

1) We're still going through some API churn so it's an ok time to make a change
2) If you really do want None I think it's easier to spell than spelling Auto which I think lots of people will want. See below.

If default is Auto and you want None:

```
x_range = Range1d(start=1, end=2, start_bound=None, end_bound=None)
```

If default is None and you want Auto:
```python
from bokeh.properties import Auto
x_range = Range1d(start=1, end=2, start_bound=Auto, end_bound=Auto)
```

Comment (bryevdv) [inline bokeh/models/ranges.py]: `Auto` is the property, a user would to this:
```
x_range = Range1d(start=1, end=2, start_bound="auto", end_bound="auto")
```

Comment (bryevdv) [inline bokeh/models/ranges.py]: Tho more importantly, +1 to making `"auto"` the default.

Comment (birdsarah) [inline bokeh/models/ranges.py]: >  I had in mind xlim behavior, i.e., factors_bound=['fat', 'cat', 'eat'] shows just those and nothing more. I imagine this kind of bound could be useful in a callback to subset the data displayed in a categorical plot without having to reload all (or any) of it?

Great - makes sense.

Comment (birdsarah) [inline bokeh/models/ranges.py]: > `x_range = Range1d(start=1, end=2, start_bound="auto", end_bound="auto")`

I see and that would still work in the case of factors, because if auto was a factor it would be factors_bound=["auto"].

Comment (birdsarah) [inline bokeh/models/ranges.py]: > I'm not sure about `start_bound` and `end_bound`. Remember, it is possible for `start > end` (i.e., if you want "reversed" range displayed) so I think `min` and `max` more clearly express the intent. Then the internals can sort out if `start` or `end` is the min, and do whatever is appropriate. Otherwise we force this on users to know and do the right thing and that seems fragile. 

I spelled it out

```python
# start < end
Range1d(start=0, end=1, start_bound=0.1, end_bound=0.9) # fine
Range1d(start=0, end=1, min=0.1, max=0.9) # fine

Range1d(start=0, end=1, start_bound=1.1, end_bound=-0.1) # error?
Range1d(start=0, end=1, min=1.1, max=-0.1) # error?

# start > end
Range1d(start=1, end=0, start_bound=1.1, end_bound=-0.1) # fine
Range1d(start=1, end=0, min=0.1, max=1.1) # fine
Range1d(start=1, end=0, max=1.1, min=0.1) # fine

Range1d(start=1, end=0, start_bound=0.1, end_bound=1.1) # error?
Range1d(start=1, end=0, min=1.1, max=0.1) # error?
```
I don't like any of these. 

New proposal:
 * Range1d
   * start - Either(Float, Datetime, Int)
   * end - Either(Float, Datetime, Int)
   * bounds - Either(Tuple(Float, Float), Tuple(Datetime, Datetime), Tuple(Int, Int), Auto, None, default=Auto)
 * DataRange1d
  * start - Float
  * end - Float
  * bounds - Either(Tuple(Float, Float), Auto, None, default=Auto)
 * FactorRange
  * factors = Either(List(String), List(Int))
  * bounds = Either(List(String), List(Int), Auto, None, default=Auto)

This gives the intuition of bounds being the same on each range so its easy to remember and lets them context switch as necessary.

Also has the advantage that if you're doing Range1d shorthand it should still be easy `Range1d(1, 0, bounds=(-0.1, 1.2))

For the Tuples, two options:
* the first value is the min and the second value is the max.
* order is irrelevant, we figure it out, larger is the max, smaller is the min.

Comment (bryevdv) [inline bokeh/models/ranges.py]: This could work, except that now users are always required to supply both bounds what if they only want a bound below? Could potentially use ellipsis for that: `bounds = (..., 10)` it's slightly unusual but not terribly so. I'd keep to mathematical convention and have it always be `(min, max)`.  There is also some slight discrepancy in that in mathematics `(0,10)` and `[0,10]` are not technically the same, but we would treat all intervals as closed (the latter).

Another nice thing: I was just thinking today that it would be nice to be able to easily hook up an axis bounds to a data range bounds and this would make that simple. 

Last comment, `Either(Tuple(Float, Float), Tuple(Datetime, Datetime), Tuple(Int, Int), Auto, None, default=Auto)` is alot, we might want to make a new property or maybe even a model for `Bounds`.

Comment (birdsarah) [inline bokeh/models/ranges.py]: > except that now users are always required to supply both bounds what if they only want a bound below

Good point, my vote for that would be to keep to the `None` convention meaning no bound - so `bounds=(None, 10)` in this case its imperative that first is min, second is max so we should definitely use that convention as you suggest.

If we use `None` as I suggest above, spelling all that out will be:
`Either(Tuple(Either(Float, None), Either(Float, None)), Tuple(Either(Datetime, None), Either(Datetime, None)), Tuple(Either(Int, None), Either(Int, None)), Auto, None, default=Auto)`
which is definitely a lot so I guess we'll definitely need a new property or model for `Bounds` as you suggest, welcome a nudge on which you think is best.

Comment (bryevdv) [inline bokeh/models/ranges.py]: I was really trying to find a reason to use `Ellipsis` for the first time in my career. :)

I think properties will actually accept `None` implicitly for most anything, so that level of specificity may not be needed unless we decide to tighten up that particular weirdness.

Comment (pzwang) [inline bokeh/models/ranges.py]: What is the way to distinguish between "tight" bounds policy and "nice fit" bounds policy?  (Recall that "nice fit" is actually contingent on screen space transforms, because what's nice for a linear scale is not the same as what's nice for a log scale.)

Comment (birdsarah) [inline bokeh/models/ranges.py]: Good question. Is that something we could address in a follow-on PR? At the moment, I believe this implementation gives you tight.

Comment (bryevdv) [inline bokeh/models/ranges.py]: Data ranges interrogate renderers for their preferred bounds, so I think its actually the converse. The current behavior is to try and give "nice" bounds. Although, its dependent on the quality of the bounding box returned by the renderer, and not all of them do an equally good job. 

"Strict" bounds is just a simple enveloping of all the data in some set of columns in a data source or sources. It's the way column data sources originally behaved, and the reason for hacks like `range_padding`. (Sometimes "range padding is useful, but it's not sufficient to implement "nice" bounds and thats what it what it was trying to be used for.) This behavior was replaced with the "ask the renderer" behavior some time ago. 

We could certainly add back a range that does simple enveloping, and then maybe an `AutoRange` that has a policy that delegates to one of the specific types (I would not want to dump the strict enveloping code back into the current data range)

Comment (bryevdv) [inline bokeh/models/ranges.py]: I also just want to say: I knew this discussion would spiral out. :) All this has happened before, and will happen again...

This will be a vey nice feature to add though, thanks for pressing on it and working through all the details @birdsarah

Comment (birdsarah) [inline bokeh/models/ranges.py]: I have added a new issue for the bounds policy - #3239.

I am going to implement `bounds` as we've discussed.

Comment (bryevdv) [inline bokeh/models/ranges.py]: We've discussed alot, can you summarize the plan? Specifically for half-bounded cases, but a full overview would get everyone on the same page.

**Comment by birdsarah** [comment]
> We've discussed alot, can you summarize the plan? Specifically for half-bounded cases, but a full overview would get everyone on the same page. 

@bryevdv apologies, I didn't see this message before I started implementing. Hopefully this sums it up: https://github.com/bokeh/bokeh/blob/feature/geo/zoom_max_min/bokeh/models/ranges.py#L47

````python
    bounds = Either(
        Auto,
        Tuple(Float, Float),
        Tuple(Datetime, Datetime),
        Tuple(Int, Int),
        help="""
    The bounds that the range is allowed to go to - typically used to prevent
    the user from panning/zooming/etc away from the data.
    By default, the bounds will be computed to the start and end of the Range.
    Bounds are provided as a tuple of ``(min, max)`` so regardless of whether your range is
    increasing or decreasing, the first item should be the minimum value of the range and the
    second item should be the maximum. Setting min > max will result in a ``ValueError``.
    Setting bounds to ``None`` will allow your plot to pan/zoom as far as you want. If you only
    want to constrain one end of the plot, you can set min or max to None.
    Examples:
        Range1d(0, 1)  # Increasing range, auto-bounded to 0 and 1 (Default behavior)
        Range1d(0, 1, bounds=(-0.1, 1.1))  # Increasing range with bounds at -0.1 and 1.1
        Range1d(1, 0, bounds=(-0.1, 1.1))  # Decreasing range with bounds at -0.1 and 1.1
        Range1d(0, 1, bounds=(0, None))  # Increasing range bounded at minimum of 0, unbounded maximum
        Range1d(start=0, end=1, bounds=None)  # Unbounded range
    """)
````

Given that I jumped the gun, I'll finish implementing python side and then wait for any comments before finalizing the JS side.

**Comment by birdsarah** [comment]
@bokeh/dev all the docstrings now describe intended behavior - welcome another round of feedback, will continue working on this next week.

There are a couple of additional cases that I can imagine wanting to support that I don't intend to in this PR as this is already a good leap forward. In particular on DataRange1d supporting `bounds=('auto', None)` 

But its probably worth pushing out this initial implementation and then seeing what people ask for.

**Comment by birdsarah** [comment]
(it's not ready for merge, but is ready for feedback - marking it as such to get feedback)

**Comment by birdsarah** [comment]
I intend to add a few more integration tests, but this is ready for review @bokeh/dev

**Comment by birdsarah** [comment]
Alright, reversed ranges are now supported and I've added the tests I wanted to, so this should be good.

**Comment by fpliger** [comment]
LGTM
